### ISSUE TITLE
fix REMOVE processing in eudev

### ISSUE THREAD

**Issue by chris2553** [gentoo/eudev#119]
I've recently built eudev-3.1.1 from the source tarball and found that it has a minor bug that was fixed in systemd after the code cut was vut.

The bug is to do with when devices are removed. I don't know the full effect of the bug, but for sure one problem was that, whilst the device nodes in /dev where deleted, any symlinks that were created when the device was added were not deleted on removal of the device, leaving dangling symlinks.

The description of the patch applied by the systemd project is:
"This reverts b67f944. Lazy loading of device properties does not work for devices
that are received over netlink, as these are sealed. Reinstate the unconditional
loading of the device db."

You can view their patch at
http://cgit.freedesktop.org/systemd/systemd/patch/src/udev/udev-event.c?id=107f2e2526d476c6cc9b81a690391c111027d641

Of course, with systemd's udev being dependent on code in systemd itself, my patch (below) is not the exact fix that was applied to their code by the systemd project. From the testing I've done here, however, my fix works and I haven't come across any regressions.

-- eudev-3.1.1/src/udev/udev-event.c~	2015-06-18 07:25:57.638462427 +0100
+++ eudev-3.1.1/src/udev/udev-event.c	2015-06-18 07:56:21.602425207 +0100
@@ -811,6 +811,10 @@ void udev_event_execute_rules(struct ude
                 return;
 
         if (streq(udev_device_get_action(dev), "remove")) {
+                udev_device_read_db(dev);
+                udev_device_tag_index(dev, NULL, false);
+		udev_device_delete_db(dev);
+
                 if (major(udev_device_get_devnum(dev)) != 0)
                         udev_watch_end(event->udev, dev);
 
@@ -821,9 +825,6 @@ void udev_event_execute_rules(struct ude
 
                 if (major(udev_device_get_devnum(dev)) != 0)
                         udev_node_remove(dev);
-
-                udev_device_delete_db(dev);
-                udev_device_tag_index(dev, NULL, false);
         } else {
                 event->dev_db = udev_device_clone_with_db(dev);
                 if (event->dev_db != NULL) {
--- eudev-3.1.1/src/libudev/libudev.h~	2015-06-18 08:07:21.474411742 +0100
+++ eudev-3.1.1/src/libudev/libudev.h	2015-06-18 08:10:15.450408192 +0100
@@ -88,6 +89,7 @@ struct udev_device *udev_device_get_pare
 struct udev_device *udev_device_get_parent_with_subsystem_devtype(struct udev_device *udev_device,
                                                                   const char *subsystem, const char *devtype);
 /* retrieve device properties */
+int udev_device_read_db(struct udev_device *udev_device);
 const char *udev_device_get_devpath(struct udev_device *udev_device);
 const char *udev_device_get_subsystem(struct udev_device *udev_device);
 const char *udev_device_get_devtype(struct udev_device *udev_device);
--- eudev-3.1.1/src/libudev/libudev-device.c~	2015-06-18 08:05:32.494413966 +0100
+++ eudev-3.1.1/src/libudev/libudev-device.c	2015-06-18 08:06:18.390413029 +0100
@@ -37,7 +37,6 @@
 #include "libudev-private.h"
 
 static int udev_device_read_uevent_file(struct udev_device *udev_device);
-static int udev_device_read_db(struct udev_device *udev_device);
 static int udev_device_set_devnode(struct udev_device *udev_device, const char *devnode);
 static struct udev_list_entry *udev_device_add_property_internal(struct udev_device *udev_device, const char *key, const char *value);
 
@@ -594,7 +593,7 @@ _public_ const char *udev_device_get_pro
         return udev_list_entry_get_value(list_entry);
 }
 
-static int udev_device_read_db(struct udev_device *udev_device)
+int udev_device_read_db(struct udev_device *udev_device)
 {
         char filename[UTIL_PATH_SIZE];
         char line[UTIL_LINE_SIZE];

**Comment by blueness** [comment]
Ouch.  That patch is almost impossible to read just pasted like that.  Take a look at "mastering markdown" for how you can make sure everything is nicely formatted.  Also read up on pull request.

Anyhow thanks for the patch.   It is correct and will be out in 3.1.2.  Please test.  Also, can you give me a use case that demonstrates the  dangling sym link problem.  How did you hit it?

**Comment by chris2553** [comment]
On 06/18/15 18:28, blueness wrote:
> Ouch. That patch is almost impossible to read just pasted like that. Take a look at "mastering markdown" for how you can
> make sure everything is nicely formatted. Also read up on pull request

I'm sorry, but it's getting more difficult by the day for an ordinary Linux user like myself to give a little back to
developers by identifying fixes for problems I encounter and sending them on to the project. When I first started using
Linux-based systems for my home computing about 15 years ago, developers simply took patches and passed back their
thanks. Over the years, I've provided hundreds of patches, mostly trivial, some not so. Nowadays, a user who tries to
give something back seems to be expected to learn to use every development tool under the sun. I provided a few FTBFS
fixes for a well known project a couple of weeks ago and was moaned at consistently because, essentially, I am not an
expert in using git! I won't be sending that project any more patches - I'll just carry them locally until they identify
the bugs and implement the related fixes. To provide you with a pull request, it seems I would have to fork eudev. I've
no intention of doing that. In all likelihood, I'd never use it again and it would just be more clutter on github. Moan
over - sorry!

> Anyhow thanks for the patch. It is correct and will be out in 3.1.2. Please test. Also, can you give me a use case
that demonstrates the dangling sym link problem. How did you hit it?

I use some local udev rules to get consistent names for the partitions on plug-in (usb-storage) devices and, as you
would expect, the rules include the SYMLINK directive. That's not unusual, of course. What I also have is a small
application running in background that watches for creation and deletion of these symlinks in /dev. On creation of a
symlink, the application looks for a hidden <symlink>.desktop file in my $HOME/DESKTOP directory and if it exists,
unhides it, so that an icon for the device pops up on my desktop. On removal of the device (and consequently, the
symlink) the application hides the .desktop file again, so that the icon disappears from my desktop. With the bug in
eudev-3.1.1, that latter (hiding) part didn't work. <symlink> is also the name of the directory in /media where the
device gets mounted (via entries in /etc/fstab) when I double-click on the icon.

I accept that this home-brewed solution to "desktop management" might be unique to my system, but I can well imagine
other users having scripts that rely on the presence of these (naming-consistency) symlinks to ascertain whether or not
a particular device is present.
